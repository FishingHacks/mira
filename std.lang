extern fn alloc(size: usize) -> &void;
extern fn realloc(ptr: &void, size: usize) -> &void;
extern fn free(ptr: &void);

@lang("allocator_trait")
trait Allocator {
    fn alloc(self: &Self, size: usize) -> &void;
    fn realloc(self: &Self, ptr: &void, size: usize) -> &void;
    fn free(self: &Self, ptr: &void);
}

struct CAlloc {;
    impl Allocator {
        fn alloc(self: &Self, size: usize) -> &void = alloc(size);
        fn realloc(self: &Self, ptr: &void, size: usize) -> &void = realloc(ptr, size);
        fn free(self: &Self, ptr: &void) = free(ptr);
    }
}
@lang("allocator")
let allocator: CAlloc = CAlloc {};

@lang("copy_trait")
trait Copy {}

@lang("clone_trait")
trait Clone {
    fn clone(self: &Self) -> Self;
}

@lang("printf")
extern fn printf(pat: &u8, a: usize, b: &u8);
