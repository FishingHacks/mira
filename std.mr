extern fn malloc(size: usize) -> &void;
extern fn realloc(ptr: &void, size: usize) -> &void;
extern fn free(ptr: &void);

@lang("allocator_trait")
trait Allocator {
    fn alloc(self: &Self, size: usize) -> &void;
    fn realloc(self: &Self, ptr: &void, size: usize) -> &void;
    fn free(self: &Self, ptr: &void);
}

struct CAlloc {;
    impl Allocator {
        fn alloc(self: &Self, size: usize) -> &void = malloc(size);
        fn realloc(self: &Self, ptr: &void, size: usize) -> &void = realloc(ptr, size);
        fn free(self: &Self, ptr: &void) = free(ptr);
    }
}
@lang("allocator")
let allocator: CAlloc = CAlloc {};

@lang("copy_trait")
trait Copy {}

@lang("clone_trait")
trait Clone {
    fn clone(self: &Self) -> Self;
}

@lang("printf")
@ext_vararg()
extern fn printf(pat: &u8);

@lang("eq_trait")
trait Eq {
    fn eq(self: &Self, other: &Self) -> bool;
}

@lang("neq_trait")
trait Neq {
    fn neq(self: &Self, other: &Self) -> bool;
}

struct Test2 {
    test: &Test
}

struct Test {
    value: i32
}

struct Nya {
    meow: i32,
    mreow: u32,
    boop: usize;

    fn new(meow: i32, mreow: u32, boop: usize) -> Nya = Nya { meow: meow, mreow: mreow, boop: boop };
}

let meow: i32 = 12;
let meowmeow: &str = "fdkjghdafhgjkdsfg";

fn test() -> Nya = Nya::new(13, 14, 15);
fn test2() -> Nya = Nya { meow: 13, mreow: 3948, boop: 0xff87cd };
fn test3(v: &Nya) -> &u32 = &v.mreow;
fn test4(v: [Nya; 5], idx: usize) -> Nya = v[idx];
fn test5(v: &[Nya]) -> usize = v[2].boop;

fn test6() -> &[&u8] = &[&3, &4];
fn test7(v: &[&[u8]], a: usize, b: usize) -> u8 = v[a][b];
fn test8() -> &str = "abcdefg :3";
fn test9() -> i32 = woof(12);

extern fn main(argc: usize, argv: &&u8) {
    let v = [test2(), test2(), test2(), test2(), test2()];
    // turn argv of &&u8 into &&str into &[&str] where the size is argc
    printf("got %d arguments\n\0" as &u8, argc);
    let value = test9();
    printf("value of test9: %d\n\0" as &u8, value);
    let value2 = test4(v, 3);
    let value3 = test3(&value2);
    printf("value of test 3 and test 4 is: %d\n\0" as &u8, *value3);
}

use "test"::woof;
