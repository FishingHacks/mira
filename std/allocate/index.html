<!DOCTYPE html><html lang="en" color-scheme="preference"><head><title>allocate - mira</title><script>const root = "../../";</script><link rel="stylesheet" href="../../index.css"><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><script src="../../index.js"></script></head><body><div class="sidebar"><h2>Packages</h2><a href="../../std/index.html">std</a></div><div class="content"><div class="search-header"><input class="search-input" placeholder="press s or / to start searching" /><a class="settings-button">Settings</a><a class="expand-button open">Summary</a><div class="settings-popup hidden"><div>Theme</div><label for="theme-light" class="setting-radio"><input type="radio" name="theme" id="theme-light" value="light"><span>light</span></label><label for="theme-dark" class="setting-radio"><input type="radio" name="theme" id="theme-dark" value="dark"><span>dark</span></label><label for="theme-ayu" class="setting-radio"><input type="radio" name="theme" id="theme-ayu" value="ayu"><span>ayu</span></label><label for="theme-preference" class="setting-radio"><input type="radio" name="theme" id="theme-preference" value="preference"><span>system preference</span></label></div></div><div class="search-results hidden"></div><div class="main-content"><div class="path"><a href="../../std/index.html">std</a></div><h1>module <span class="module">allocate</span></h1><details class="description-toggle" open><summary><span>Expand</span></summary><div class="description"><p>Module responsible for handling allocations</p>
</div></details><h2 id="structs" class="anchorable header">Structs<a class="anchor" href="#structs">ยง</a></h2><dl class="item-list"><dt><a href="../../std/allocate/CAlloc.html" class="item-ref struct">CAlloc</a></dt><dd class="description"><p>The standard libc allocator (currently only returns a null pointer.)</p>
</dd></dl><h2 id="traits" class="anchorable header">Traits<a class="anchor" href="#traits">ยง</a></h2><dl class="item-list"><dt><a href="../../std/allocate/Allocator.html" class="item-ref trait">Allocator</a></dt><dd class="description"><p>Some struct that allocates memory.</p>
</dd></dl><h2 id="statics" class="anchorable header">Statics<a class="anchor" href="#statics">ยง</a></h2><dl class="item-list"><dt><a href="../../std/allocate/allocator.html" class="item-ref static">allocator</a></dt><dd class="description"><p>The global allocator.</p>
</dd></dl></div></div></body></html>