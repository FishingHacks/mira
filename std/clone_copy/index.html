<!DOCTYPE html><html lang="en" color-scheme="preference"><head><title>clone_copy - mira</title><script>const root = "../../";</script><link rel="stylesheet" href="../../index.css"><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><script src="../../index.js"></script></head><body><div class="sidebar"><h2>Packages</h2><a href="../../std/index.html">std</a></div><div class="content"><div class="search-header"><input class="search-input" placeholder="press s or / to start searching" /><a class="settings-button">Settings</a><div class="settings-popup hidden"><div>Theme</div><label for="theme-light" class="setting-radio"><input type="radio" name="theme" id="theme-light" value="light"><span>light</span></label><label for="theme-dark" class="setting-radio"><input type="radio" name="theme" id="theme-dark" value="dark"><span>dark</span></label><label for="theme-ayu" class="setting-radio"><input type="radio" name="theme" id="theme-ayu" value="ayu"><span>ayu</span></label><label for="theme-preference" class="setting-radio"><input type="radio" name="theme" id="theme-preference" value="preference"><span>system preference</span></label></div></div><div class="search-results hidden"></div><div class="main-content"><div class="path"><a href="../../std/index.html">std</a></div><h1>module <span class="module">clone_copy</span></h1><details class="description-toggle" open><summary><span>Expand</span></summary><div class="description"><p>Module holding the <code>Clone</code> and <code>Copy</code> traits</p>
</div></details><h2 id="traits" class="anchorable header">Traits<a class="anchor" href="#traits">ยง</a></h2><dl class="item-list"><dt><a href="../../std/clone_copy/Clone.html" class="item-ref trait">Clone</a></dt><dd class="description"><p>Gives the ability for the user to Clone the type.</p>
</dd><dt><a href="../../std/clone_copy/Copy.html" class="item-ref trait">Copy</a></dt><dd class="description"><p>Currently does nothing.</p>
</dd></dl></div></div></body></html>