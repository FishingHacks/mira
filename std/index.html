<!DOCTYPE html><html lang="en" color-scheme="preference"><head><title>std - mira</title><script>const root = "../";</script><link rel="stylesheet" href="../index.css"><noscript><link rel="stylesheet" href="../noscript.css"></noscript><script src="../index.js"></script></head><body><div class="sidebar"><h2>Packages</h2><a href="../std/index.html">std</a></div><div class="content"><div class="search-header"><input class="search-input" placeholder="press s or / to start searching" /><a class="settings-button">Settings</a><div class="settings-popup hidden"><div>Theme</div><label for="theme-light" class="setting-radio"><input type="radio" name="theme" id="theme-light" value="light"><span>light</span></label><label for="theme-dark" class="setting-radio"><input type="radio" name="theme" id="theme-dark" value="dark"><span>dark</span></label><label for="theme-ayu" class="setting-radio"><input type="radio" name="theme" id="theme-ayu" value="ayu"><span>ayu</span></label><label for="theme-preference" class="setting-radio"><input type="radio" name="theme" id="theme-preference" value="preference"><span>system preference</span></label></div></div><div class="search-results hidden"></div><div class="main-content"><div class="path"></div><h1>module <span class="module">std</span></h1><div class="description"><h1>Mira Standard Library</h1>
<p>Standard library for mira. Here is where all the intrinsics and lang items live too. (compiler internals yippieeee)</p>
</div><h2 id="modules" class="anchorable header">Modules<a class="anchor" href="#modules">§</a></h2><dl class="item-list"><dt><a href="../std/langitems/index.html" class="item-ref module">langitems</a></dt><dd class="description"><p>Lang items</p>
</dd><dt><a href="../std/intrinsics/index.html" class="item-ref module">intrinsics</a></dt><dd class="description"><p>Compiler intrinsics – you really shouldn’t touch this, unless some intrinsic is exported to somewhere else.</p>
</dd><dt><a href="../std/allocate/index.html" class="item-ref module">allocate</a></dt><dd class="description"><p>Module responsible for handling allocations</p>
</dd><dt><a href="../std/os/index.html" class="item-ref module">os</a></dt><dd class="description"><p>Operating System Interactions</p>
</dd><dt><a href="../std/panicking/index.html" class="item-ref module">panicking</a></dt><dd class="description"><p>Panicking</p>
</dd><dt><a href="../std/clone_copy/index.html" class="item-ref module">clone_copy</a></dt><dd class="description"><p>Module holding the <code>Clone</code> and <code>Copy</code> traits</p>
</dd></dl><h2 id="functions" class="anchorable header">Functions<a class="anchor" href="#functions">§</a></h2><dl class="item-list"><dt><a href="../std/print_str_newline.html" class="item-ref function">print_str_newline</a></dt><dd class="description"><p>Prints <code>string</code> with a newline (<code>\n</code>) to stdout</p>
</dd><dt><a href="../std/cstrlen.html" class="item-ref function">cstrlen</a></dt><dd class="description"><p>Gets the length of a c string</p>
</dd><dt><a href="../std/panicking/panic.html" class="item-ref function">panic</a></dt><dd class="description"><p>Prints “panic: {message}\n”, prints the backtrace and exits.</p>
</dd><dt><a href="../std/print_newline.html" class="item-ref function">print_newline</a></dt><dd class="description"><p>Prints a newliine to stdout</p>
</dd><dt><a href="../std/langitems/print_u64.html" class="item-ref function">print_u64</a></dt><dd class="description"><p>Prints a <code>u64</code> to stdout</p>
</dd><dt><a href="../std/langitems/print_i64.html" class="item-ref function">print_i64</a></dt><dd class="description"><p>Prints an <code>i64</code> to stdout</p>
</dd><dt><a href="../std/print_u64_hex.html" class="item-ref function">print_u64_hex</a></dt><dd class="description"><p>Prints a u64 as a hex literal</p>
</dd><dt><a href="../std/print_char.html" class="item-ref function">print_char</a></dt><dd class="description"><p>Prints the character <code>c</code> to stdout</p>
</dd><dt><a href="../std/print_str.html" class="item-ref function">print_str</a></dt><dd class="description"><p>Prints <code>string</code> to stdout</p>
</dd></dl><h2 id="traits" class="anchorable header">Traits<a class="anchor" href="#traits">§</a></h2><dl class="item-list"><dt><a href="../std/clone_copy/Copy.html" class="item-ref trait">Copy</a></dt><dd class="description"><p>Currently does nothing.</p>
</dd><dt><a href="../std/clone_copy/Clone.html" class="item-ref trait">Clone</a></dt><dd class="description"><p>Gives the ability for the user to Clone the type.</p>
</dd></dl></div></div></body></html>