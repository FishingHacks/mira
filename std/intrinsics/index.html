<!DOCTYPE html><html lang="en" color-scheme="preference"><head><title>intrinsics - mira</title><script>const root = "../../";</script><link rel="stylesheet" href="../../index.css"><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><script src="../../index.js"></script></head><body><div class="sidebar"><h2>Packages</h2><a href="../../std/index.html">std</a></div><div class="content"><div class="search-header"><input class="search-input" placeholder="press s or / to start searching" /><a class="settings-button">Settings</a><div class="settings-popup hidden"><div>Theme</div><label for="theme-light" class="setting-radio"><input type="radio" name="theme" id="theme-light" value="light"><span>light</span></label><label for="theme-dark" class="setting-radio"><input type="radio" name="theme" id="theme-dark" value="dark"><span>dark</span></label><label for="theme-ayu" class="setting-radio"><input type="radio" name="theme" id="theme-ayu" value="ayu"><span>ayu</span></label><label for="theme-preference" class="setting-radio"><input type="radio" name="theme" id="theme-preference" value="preference"><span>system preference</span></label></div></div><div class="search-results hidden"></div><div class="main-content"><div class="path"><a href="../../std/index.html">std</a></div><h1>module <span class="module">intrinsics</span></h1><details class="description-toggle" open><summary><span>Expand</span></summary><div class="description"><p>Compiler intrinsics – you really shouldn’t touch this, unless some intrinsic is exported to somewhere else.</p>
</div></details><h2 id="functions" class="anchorable header">Functions<a class="anchor" href="#functions">§</a></h2><dl class="item-list"><dt><a href="../../std/intrinsics/unreachable.html" class="item-ref function">unreachable</a></dt><dd class="description"><p>Marks an unreachable branch. Triggers UB when reached.</p>
</dd><dt><a href="../../std/intrinsics/type_name.html" class="item-ref function">type_name</a></dt><dd class="description"><p>Returns the typename of the type <code>T</code></p>
</dd><dt><a href="../../std/intrinsics/return_address.html" class="item-ref function">return_address</a></dt><dd class="description"><p>Returns the return address of the current function, or one of it’s callers, depending on if it was inlined or not.</p>
</dd><dt><a href="../../std/intrinsics/call_main.html" class="item-ref function">call_main</a></dt><dd class="description"><p>Calls the main function, if it exists</p>
</dd><dt><a href="../../std/intrinsics/get_metadata.html" class="item-ref function">get_metadata</a></dt><dd class="description"><p>Gets the metadata of the reference. This is the vtable pointer for <code>&amp;dyn _</code>, or the array length for <code>&amp;[_]</code>. Returns 0 for all sized types.</p>
</dd><dt><a href="../../std/intrinsics/abort.html" class="item-ref function">abort</a></dt><dd class="description"><p>Traps the processor, or, if no such instruction exists, aborts the process.</p>
</dd><dt><a href="../../std/intrinsics/size_of.html" class="item-ref function">size_of</a></dt><dd class="description"><p>Returns the size of the type <code>T</code></p>
</dd><dt><a href="../../std/intrinsics/size_of_val.html" class="item-ref function">size_of_val</a></dt><dd class="description"><p>Gets the size of the value of type T at the pointer. This takes into account number of elements for <code>&amp;[_]</code> or the size of the underlying type for <code>&amp;dyn _</code>.</p>
</dd><dt><a href="../../std/intrinsics/breakpoint.html" class="item-ref function">breakpoint</a></dt><dd class="description"><p>Triggers a breakpoint</p>
</dd></dl></div></div></body></html>