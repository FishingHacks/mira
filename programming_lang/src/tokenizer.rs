use std::fmt::{Display, Write};

use crate::{
    error::ProgrammingLangTokenizationError,
    globals::GlobalString,
    parser::{LiteralValue, Parser},
};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum TokenType {
    // Tokenization, Parsing ; *: no types
    Let,                     // done, done
    Const,                   // done, done
    Equals,                  // done, done
    NotEquals,               // done, done
    LessThan,                // done, done
    GreaterThan,             // done, done
    LessThanEquals,          // done, done
    GreaterThanEquals,       // done, done
    LogicalNot,              // done, done
    LogicalAnd,              // done, done
    LogicalOr,               // done, done
    StringLiteral,           // done, done
    NumberLiteral,           // done, done
    BooleanLiteral,          // done, done
    NullLiteral,             // done, done
    IdentifierLiteral,       // done, done
    AssignValue,             // done, done
    AssignTypeOrStructValue, // done, done*
    Semicolon,               // done, done
    ParenLeft,               // done, done
    ParenRight,              // done, done
    CurlyLeft,               // done, done
    CurlyRight,              // done, done
    BracketLeft,             // done, done
    BracketRight,            // done, done
    Plus,                    // done, done
    Minus,                   // done, done
    MultiplyOrDeref,         // done, done
    Divide,                  // done, done
    Modulo,                  // done, done
    ModuloAssign,            // done, done
    IntegerDivide,           // done, done
    BitwiseNot,              // done, done
    BitwiseAndOrReference,   // done, done
    BitwiseOr,               // done, done
    BitwiseXor,              // done, done
    Return,                  // done, done
    Fn,                      // done, done
    Extern,
    If,                      // done, done
    Else,                    // done, done
    While,                   // done, done
    For,                     // done, done
    In,                      // done, done
    Range,                   // done, done
    RangeInclusive,          // done, done
    ReturnType,              // done, *
    Struct,                  // done,
    Trait,                   // done, *
    Impl,                    // done,
    Comma,                   // done, done
    PlusAssign,              // done, done
    MinusAssign,             // done, done
    DivideAssign,            // done, done
    MultiplyAssign,          // done, done
    BitwiseAndAssign,        // done, done
    BitwiseOrAssign,         // done, done
    BitwiseXorAssign,        // done, done
    Dot,                     // done, done
    As,                      // done, *
    QuestionMark,            // done, *
    Eof,                     // done, done
}

#[derive(Clone, Debug)]
pub enum Literal {
    Number(f64),
    String(Box<str>),
    Null,
    Bool(bool),
}

#[derive(Debug, Clone, Copy)]
pub struct Location {
    pub line: usize,
    pub file: GlobalString,
    pub column: usize,
}

impl Location {
    pub fn new(file: GlobalString, line: usize, column: usize) -> Self {
        Self { column, file, line }
    }
}

impl Display for Location {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.file, f)?;
        f.write_char(':')?;
        Display::fmt(&self.line, f)?;
        f.write_char(':')?;
        Display::fmt(&self.column, f)
    }
}

#[macro_export]
macro_rules! loc {
    ($file:expr;$line:expr) => {
        $crate::tokenizer::Location::new($file, $line, 0)
    };
    ($file:expr;$line:expr;$column:expr) => {
        $crate::tokenizer::Location::new($file, $line, $column)
    };
}
pub(crate) use loc;

#[derive(Clone, Debug)]
pub struct Token {
    pub typ: TokenType,
    pub literal: Option<Literal>,
    pub location: Location,
}

impl Token {
    pub fn new(
        typ: TokenType,
        literal: Option<Literal>,
        line_number: usize,
        file: GlobalString,
    ) -> Self {
        Self {
            typ,
            location: loc!(file;line_number),
            literal,
        }
    }

    pub fn to_literal_value(&self) -> Option<LiteralValue> {
        match self.typ {
            TokenType::StringLiteral
            | TokenType::NullLiteral
            | TokenType::BooleanLiteral
            | TokenType::NumberLiteral => self.literal.as_ref().map(|v| match v {
                Literal::Null => LiteralValue::Null,
                Literal::Bool(b) => LiteralValue::Bool(*b),
                Literal::Number(n) => LiteralValue::Number(*n),
                Literal::String(s) => LiteralValue::String(s.clone()),
            }),
            TokenType::IdentifierLiteral => {
                if let Some(lit) = &self.literal {
                    return match &lit {
                        Literal::String(v) => Some(LiteralValue::Dynamic(v.clone())),
                        _ => None,
                    };
                } else {
                    return None;
                }
            }
            _ => None,
        }
    }
}

pub struct Tokenizer {
    source: Vec<char>,
    file: GlobalString,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
}

impl Tokenizer {
    pub fn new<I: Into<GlobalString>>(source: &str, file: I) -> Self {
        Self {
            source: source.chars().collect(),
            file: file.into(),
            start: 0,
            current: 0,
            tokens: vec![],
            line: 0,
        }
    }

    pub fn scan_tokens(&mut self) -> Result<(), Vec<ProgrammingLangTokenizationError>> {
        let mut errors = vec![];
        while !self.is_at_end() {
            self.start = self.current;
            if let Err(e) = self.scan_token() {
                errors.push(e);
            }
        }

        self.tokens.push(Token {
            typ: TokenType::Eof,
            literal: None,
            location: loc!(self.file;self.line + 1),
        });

        if errors.len() > 0 {
            Err(errors)
        } else {
            Ok(())
        }
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn advance(&mut self) -> char {
        if self.is_at_end() {
            return '\0';
        }
        self.current += 1;
        if self.current > 1 && self.source[self.current - 2] == '\n' {
            self.line += 1;
        }
        self.source[self.current - 1]
    }

    fn cur_char(&self) -> char {
        if self.current < 1 {
            self.source[0] // before advance() was called, this shouldnt happen
        } else {
            self.source[self.current - 1]
        }
    }

    fn peek(&self) -> char {
        self.source.get(self.current).map(|c| *c).unwrap_or('\0')
    }

    fn if_char_advance(&mut self, character: char) -> bool {
        if self.peek() == character {
            self.advance();
            true
        } else {
            false
        }
    }

    fn scan_token(&mut self) -> Result<(), ProgrammingLangTokenizationError> {
        let c = self.advance();

        macro_rules! token {
            ($token: ident) => {
                self.add_token(TokenType::$token)
            };

            ($tokena: ident, $tokenb: ident, $char: expr) => {{
                if self.if_char_advance($char) {
                    self.add_token(TokenType::$tokenb)
                } else {
                    self.add_token(TokenType::$tokena)
                }
            }};
        }

        match c {
            '?' => token!(QuestionMark),
            '(' => token!(ParenLeft),
            ')' => token!(ParenRight),
            '{' => token!(CurlyLeft),
            '}' => token!(CurlyRight),
            '[' => token!(BracketLeft),
            ']' => token!(BracketRight),
            ',' => token!(Comma),
            '.' => {
                if self.if_char_advance('.') {
                    if self.if_char_advance('=') {
                        token!(RangeInclusive); // ..=
                    } else {
                        token!(Range); // ..
                    }
                } else {
                    if ('0'..='9').contains(&self.peek()) {
                        return self.parse_number('.');
                    } else {
                        token!(Dot) // .
                    }
                }
            }
            '+' => token!(Plus, PlusAssign, '='),
            '-' => {
                if self.if_char_advance('=') {
                    token!(MinusAssign);
                } else if self.if_char_advance('>') {
                    token!(ReturnType);
                } else {
                    token!(Minus);
                }
            }
            '/' if self.peek() != '/' => token!(Divide, DivideAssign, '='),
            '%' => token!(Modulo, ModuloAssign, '='),
            '*' => token!(MultiplyOrDeref, MultiplyAssign, '='),
            '=' => token!(AssignValue, Equals, '='),
            '<' => token!(LessThan, LessThanEquals, '='),
            '>' => token!(GreaterThan, GreaterThanEquals, '='),
            ':' => token!(AssignTypeOrStructValue),
            ';' => token!(Semicolon),
            '!' => token!(LogicalNot, NotEquals, '='),
            '~' => token!(BitwiseNot),
            '&' => {
                if self.if_char_advance('=') {
                    token!(BitwiseAndAssign);
                } else if self.if_char_advance('&') {
                    token!(LogicalAnd);
                } else {
                    token!(BitwiseAndOrReference);
                }
            }
            '|' => {
                if self.if_char_advance('=') {
                    token!(BitwiseOrAssign);
                } else if self.if_char_advance('|') {
                    token!(LogicalOr);
                } else {
                    token!(BitwiseOr);
                }
            }
            '^' => token!(BitwiseXor, BitwiseXorAssign, '='),
            ' ' | '\n' | '\r' | '\t' => (),
            '/' if self.if_char_advance('/') => {
                while !self.is_at_end() && self.advance() != '\n' {}
            }
            ('0'..='9') => return self.parse_number(c),
            '"' => return self.parse_string(),
            _ if Self::is_valid_identifier_char(c) && !matches!(c, ('0'..='9')) => {
                self.parse_identifier(c)
            }
            _ => {
                return Err(ProgrammingLangTokenizationError::unknown_token(
                    loc!(self.file;self.line),
                    c,
                ))
            }
        }
        Ok(())
    }

    fn add_token(&mut self, token: TokenType) {
        self.tokens
            .push(Token::new(token, None, self.line, self.file));
    }

    fn parse_number(&mut self, first_char: char) -> Result<(), ProgrammingLangTokenizationError> {
        let mut str = String::with_capacity(5);
        if first_char == '.' {
            str.push('0');
        }
        str.push(first_char);

        while !self.is_at_end() {
            if matches!(self.peek(), ('0'..='9')) {
                str.push(self.advance())
            } else if self.peek() == '.'
                && matches!(
                    self.source
                        .get(self.current + 1)
                        .map(|c| *c)
                        .unwrap_or('\0'),
                    ('0'..='9')
                )
            {
                str.push(self.advance());
            } else {
                break;
            }
        }

        let num = match str.parse::<f64>() {
            Ok(num) => num,
            Err(..) => {
                return Err(ProgrammingLangTokenizationError::invalid_number(
                    loc!(self.file;self.line),
                ))
            }
        };
        self.add_token_lit(TokenType::NumberLiteral, Literal::Number(num));

        Ok(())
    }

    fn parse_string(&mut self) -> Result<(), ProgrammingLangTokenizationError> {
        let mut is_backslash = false;
        let mut str = String::with_capacity(30);

        while !self.is_at_end() {
            let c = self.advance();

            if is_backslash {
                is_backslash = false;
                str.push(Self::backslash_char_to_real_char(c));
            } else if c == '\\' {
                is_backslash = true;
            } else if c == '"' {
                break;
            } else {
                str.push(c);
            }
        }
        if self.cur_char() != '"' || self.source[self.current - 2] == '\\' {
            return Err(ProgrammingLangTokenizationError::unclosed_string(
                loc!(self.file;self.line+1),
            ));
        }

        self.add_token_lit(
            TokenType::StringLiteral,
            Literal::String(str.into_boxed_str()),
        );
        return Ok(());
    }

    fn backslash_char_to_real_char(character: char) -> char {
        match character {
            'n' => '\n',
            '0' => '\0',
            'r' => '\r',
            't' => '\t',
            _ => character,
        }
    }

    fn parse_identifier(&mut self, starting_char: char) {
        let mut identifier = String::with_capacity(10);
        identifier.push(starting_char);

        while !self.is_at_end() {
            if Self::is_valid_identifier_char(self.peek()) {
                identifier.push(self.advance());
            } else {
                let next_char = self.peek();
                let char_after_next = self.source.get(self.current + 1).map(|c| *c).unwrap_or('\0');
                if next_char == ':' && char_after_next == ':' {
                    identifier.push_str("::");
                    self.advance();
                    self.advance();
                    continue;
                }
                break;
            }
        }
        match identifier.as_str() {
            "null" => return self.add_token(TokenType::NullLiteral),
            "true" => return self.add_token_lit(TokenType::BooleanLiteral, Literal::Bool(true)),
            "false" => return self.add_token_lit(TokenType::BooleanLiteral, Literal::Bool(false)),
            _ => (),
        }
        if let Some(typ) = Self::try_token_from_keyword(&identifier) {
            return self.add_token(typ);
        }
        self.add_token_lit(
            TokenType::IdentifierLiteral,
            Literal::String(identifier.into()),
        );
    }

    fn try_token_from_keyword(word: &str) -> Option<TokenType> {
        match word {
            "let" => Some(TokenType::Let),
            "const" => Some(TokenType::Const),
            "as" => Some(TokenType::As),
            "fn" => Some(TokenType::Fn),
            "extern" => Some(TokenType::Extern),
            "return" => Some(TokenType::Return),
            "if" => Some(TokenType::If),
            "else" => Some(TokenType::Else),
            "while" => Some(TokenType::While),
            "for" => Some(TokenType::For),
            "in" => Some(TokenType::In),
            "idiv" => Some(TokenType::IntegerDivide),
            "struct" => Some(TokenType::Struct),
            "impl" => Some(TokenType::Impl),
            "trait" => Some(TokenType::Trait),
            _ => None,
        }
    }

    fn add_token_lit(&mut self, token: TokenType, literal: Literal) {
        self.tokens
            .push(Token::new(token, Some(literal), self.line, self.file));
    }

    // valid characters:
    // abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_$#
    fn is_valid_identifier_char(character: char) -> bool {
        matches!(
            character,
            '_' | '$' | '#' | ('a'..='z') | ('A'..='Z') | ('0'..='9') // TODO: generics / types: make TypeName::FunctionName use some custom token as :: instead of having that part of the type
        )
    }

    pub fn get_tokens(&self) -> &Vec<Token> {
        &self.tokens
    }

    pub fn to_parser(self) -> Parser {
        Parser::new(self.tokens)
    }
}
