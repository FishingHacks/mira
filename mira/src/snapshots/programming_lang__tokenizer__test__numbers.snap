---
source: programming_lang/src/tokenizer.rs
expression: "(errs, tokens)"
---
(
    [
        text:8:4: Could not parse the number,
        text:9:5: Could not parse the number,
        text:13:5: Could not parse the number,
        text:14:6: Could not parse the number,
        text:18:5: Could not parse the number,
        text:19:6: Could not parse the number,
    ],
    [
        Token {
            typ: UIntLiteral,
            literal: Some(
                UInt(
                    12,
                ),
            ),
            location: Location {
                line: 1,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 1,
                column: 2,
                file: "text",
            },
        },
        Token {
            typ: SIntLiteral,
            literal: Some(
                SInt(
                    -23,
                ),
            ),
            location: Location {
                line: 2,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 2,
                column: 3,
                file: "text",
            },
        },
        Token {
            typ: FloatLiteral,
            literal: Some(
                Float(
                    23.9,
                ),
            ),
            location: Location {
                line: 3,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 3,
                column: 4,
                file: "text",
            },
        },
        Token {
            typ: FloatLiteral,
            literal: Some(
                Float(
                    -29.3,
                ),
            ),
            location: Location {
                line: 4,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 4,
                column: 5,
                file: "text",
            },
        },
        Token {
            typ: UIntLiteral,
            literal: Some(
                UInt(
                    1,
                ),
            ),
            location: Location {
                line: 6,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 6,
                column: 3,
                file: "text",
            },
        },
        Token {
            typ: SIntLiteral,
            literal: Some(
                SInt(
                    -1,
                ),
            ),
            location: Location {
                line: 7,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 7,
                column: 4,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 8,
                column: 5,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 9,
                column: 6,
                file: "text",
            },
        },
        Token {
            typ: UIntLiteral,
            literal: Some(
                UInt(
                    13,
                ),
            ),
            location: Location {
                line: 11,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 11,
                column: 6,
                file: "text",
            },
        },
        Token {
            typ: SIntLiteral,
            literal: Some(
                SInt(
                    -5,
                ),
            ),
            location: Location {
                line: 12,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 12,
                column: 6,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 13,
                column: 6,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 14,
                column: 7,
                file: "text",
            },
        },
        Token {
            typ: UIntLiteral,
            literal: Some(
                UInt(
                    5,
                ),
            ),
            location: Location {
                line: 16,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 16,
                column: 3,
                file: "text",
            },
        },
        Token {
            typ: SIntLiteral,
            literal: Some(
                SInt(
                    -34,
                ),
            ),
            location: Location {
                line: 17,
                column: 0,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 17,
                column: 5,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 18,
                column: 6,
                file: "text",
            },
        },
        Token {
            typ: Semicolon,
            literal: None,
            location: Location {
                line: 19,
                column: 7,
                file: "text",
            },
        },
        Token {
            typ: Eof,
            literal: None,
            location: Location {
                line: 20,
                column: 0,
                file: "text",
            },
        },
    ],
)
